Yosemite/Yosemite/Stores/AppSettingsStore.swift

```diff
@@ -0,0 +1,171 @@
+    ///
+    private let fileStorage: FileStorage
+
+    /// Designated initaliser
+    ///
+    public init(dispatcher: Dispatcher,
+                storageManager: StorageManagerType,
+                fileStorage: FileStorage) {
+        self.fileStorage = fileStorage
+        super.init(dispatcher: dispatcher,
+                   storageManager: storageManager,
+                   network: NullNetwork())
+    }
+
+    /// URL to the plist file that we use to store the user selected
+    /// shipment tracing provider. Not declared as `private` so it can
+    /// be overriden in tests
+    ///
+    lazy var selectedProvidersURL: URL = {
+        let documents = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first
+        return documents!.appendingPathComponent("shipment-providers.plist")
+    }()
+
+    /// Registers for supported Actions.
+    ///
+    override public func registerSupportedActions(in dispatcher: Dispatcher) {
+        dispatcher.register(processor: self, for: AppSettingsAction.self)
+    }
+
+    /// Receives and executes Actions.
+    ///
+    override public func onAction(_ action: Action) {
+        guard let action = action as? AppSettingsAction else {
+            assertionFailure("ShipmentStore received an unsupported action")
+            return
+        }
+
+        switch action {
+        case .addTrackingProvider(let siteID, let providerName, let onCompletion):
+            addTrackingProvider(siteID: siteID,
+                                providerName: providerName,
+                                onCompletion: onCompletion)
+
+        case .loadTrackingProvider(let siteID, let onCompletion):
+            loadTrackingProvider(siteID: siteID,
+                                 onCompletion: onCompletion)
+        }
+    }
+}
+
+
+// MARK: - Shipment tracking providers!
+//
+private extension AppSettingsStore {
+    func addTrackingProvider(siteID: Int,
+                             providerName: String,
+                             onCompletion: (Error?) -> Void) {
+        guard FileManager.default.fileExists(atPath: selectedProvidersURL.path) else {
+            insertNewProvider(siteID: siteID,
+                              providerName: providerName,
+                              onCompletion: onCompletion)
+            onCompletion(nil)
+            return
+        }
+
+        do {
+            let data = try fileStorage.data(for: selectedProvidersURL)
+            let decoder = PropertyListDecoder()
+            let settings = try decoder.decode([PreselectedProvider].self, from: data)
+            upsertTrackingProvider(siteID: siteID,
+                                   providerName: providerName,
+                                   preselectedData: settings,
+                                   onCompletion: onCompletion)
+        } catch {
+            let error = AppSettingsStoreErrors.parsePreselectedProvider
+            onCompletion(error)
```

Comment:
Could we add a `DDLogError(error)` here, to give hints in the log that something went wrong?